src/devices/src/legacy/serial.rs

```diff
@@ -236,29 +246,112 @@ impl Serial {
         }
     }
 
+    #[inline]
+    fn avail_buffer_capacity(&self) -> usize {
+        FIFO_SIZE.checked_sub(self.in_buffer.len()).unwrap_or_else(||
+            panic!(
+                "Errored out due to serial device buffer size greater than the maximum expected size: {} > {}.",
+                self.in_buffer.len(),
+                FIFO_SIZE
+            )
+        )
+    }
+
     fn recv_bytes(&mut self) -> io::Result<usize> {
+        let avail_cap = self.avail_buffer_capacity();
+        if avail_cap == 0 {
+            return Err(io::Error::from_raw_os_error(libc::ENOBUFS));
+        }
+
         if let Some(input) = self.input.as_mut() {
-            let mut out = [0u8; 32];
-            return input.read(&mut out).and_then(|count| {
-                if count > 0 {
-                    self.raw_input(&out[..count])?;
-                    Ok(count)
-                } else {
-                    Ok(0)
-                }
-            });
+            let mut out = vec![0u8; avail_cap];
```

Comment:
I think this should be a member of the struct as the constant allocation and zeroing of the memory is costly, but this is an optimization which we can implement later and is not related to the security issue.